### PR TITLE
Support draft-ietf-acme-ari-03

### PR DESCRIPTION
**Pull request by pgporada**
I've updated all draft-ietf-acme-ari-01 code with the latest version, [draft-ietf-acme-ari-03](https://datatracker.ietf.org/doc/html/draft-ietf-acme-ari-03). Big thanks to @beautifulentropy for [this blog post](https://letsencrypt.org/2024/04/25/guide-to-integrating-ari-into-existing-acme-clients) which made the work very easy. Client authors can specify an optional `*x509.Certificate` for ARI certificate replacements attempts via `client.ReplacementOrder`. The existing `client.NewOrder` and `client.NewOrderDomains` become a thin wrapper around `client.ReplacementOrder` that will not trigger a certificate replacement.

### CODE REVIEW

**Comment by eggsampler** [comment]
Thanks for the pull request! I might not be able to review and pull this in until the weekend, if there is no immediate urgency.

**Comment by pgporada** [comment]
That's ok, Aaron made a good point, I still have more work to do.

File: order.go
Comment (aarongable) [inline order.go:70]: Rather than being a method with an `*Order` receiver, and modifying that receiver in-place, this should be a method on the Client which returns `(*Order, error)`.

None of the places where you call this function populate the receiver with anything before calling this method, so it's not serving a purpose. All of the
```go
newOrderResp := Order{}
err = newOrderResp.postNewOrder(...)
```
can become
```go
newOrderResp, err := c.postNewOrder(...)
```

File: order.go
Comment (aarongable) [inline order.go:43]: Is there a compelling reason to make this a separate method, rather than adding a new `replaces *Order` argument to the existing c.NewOrderDomains()?

Comment (pgporada) [reply order.go:43]: I didn't want to affect other downstream code that calls `c.NewOrderDomains`. I can change it if you want.

File: order.go
Comment (pgporada) [reply order.go:70]: There was quite a bit of surgery I had to do to pass around a pointer to an `Object`.

File: account.go
Comment (pgporada) [inline account.go:23]: My editor was throwing a warning about this so I decided to take care of it.

File: order.go
Comment (aarongable) [inline order.go:1]: Two comments on this file overall:
1) I don't see why it's necessary to change the `Order` return type to an `*Order`. I don't see any code here taking advantage of interior mutability. And changing this type breaks backwards-compatibility just as surely as adding a new "oldCert" argument to the existing functions would.
2) I still think that adding a new "oldCert" argument to the existing functions would be better. Here you're creating two new layers of abstraction: a new "NewOrderRenewal" method which operates differently than its sibling "NewOrder" and "NewOrderDomains" siblings, and the new `postNewOrder` and `domainsToIds` helpers which sit underneath those methods. If the "oldCert" argument were simply added to the existing NewOrder and NewOrderDomains methods, then NewOrder could continue to contain all of the posting logic without the need for a helper, and NewOrderDomains could continue to be a simple wrapper around NewOrder without the need for a helper. But of course, I'm not a maintainer of this library, so I defer to the maintainers' judgement here.

Comment (pgporada) [reply order.go:1]: For #1 that was the result of https://github.com/eggsampler/acme/pull/23#discussion_r1586814331.  I probably went overboard somewhere though.

Comment (eggsampler) [reply order.go:1]: Am looking into the code now. First glance agree with aarongable on both points, will make some modifications and suggestion shortly.

**Comment by eggsampler** [comment]
I've made a few updates which I think streamline the code a bit more. Specifically around streamlining the order process which removes the need for the `postNewOrder` function, and not making any changes to existing published signature functions.

Have run a few tests against boulder and looks good from my end, but I have not been following the updates in the ARI space. If you get a chance in the next few days would appreciate a quick eye over before I merge and bump the version.

**Review by aarongable — COMMENTED** [review]
LGTM, I think this works. Just two small optional comments.

**Review by aarongable — COMMENTED** [review]
LGTM, for whatever it's worth :) Thanks @pgporada for putting this together, and thanks so much to @eggsampler for reviewing, improving, and considering this contribution!

**Review by eggsampler — APPROVED** [review]
Looks good to me too, thankyou!

**Comment by pgporada** [comment]
Merge it when ready :)

**Pull request merged**
